subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentGraphResolveState.java

```diff
@@ -151,7 +151,17 @@ private class DefaultLocalConfigurationGraphResolveState extends AbstractVariant
         public DefaultLocalConfigurationGraphResolveState(long instanceId, LocalComponentMetadata component, LocalConfigurationMetadata configuration, Lazy<Optional<Set<? extends VariantResolveMetadata>>> allVariantsForArtifactSelection) {
             this.instanceId = instanceId;
             this.configuration = configuration;
-            this.artifactResolveState = Lazy.locking().of(() -> new DefaultLocalConfigurationArtifactResolveState(component, configuration, allVariantsForArtifactSelection));
+            // We deliberately avoid locking the initialization of `artifactResolveState`.
+            // This object may be shared across multiple worker threads, and the computation of
+            // `legacyVariants` below is likely to require acquiring the state lock for the
+            // project that owns this `Configuration` leading to a potential deadlock situation.
+            // For instance, a thread could acquire the `artifactResolveState` lock while another thread,
+            // which already owns the project lock, attempts to acquire the `artifactResolveState` lock.
+            // See https://github.com/gradle/gradle/issues/25416
+            this.artifactResolveState = Lazy.nonAtomic().of(() -> {
+                Set<? extends VariantResolveMetadata> legacyVariants = configuration.prepareToResolveArtifacts().getVariants();
```

Comment:
❔ Why move this outside of the constructor?  Is this just to keep the constructor logic simpler?